### PR TITLE
Python parser

### PR DESCRIPTION
**Pull request by singhsarab**
First draft for changes related to python unittest log parser

Pending:
1. More Unit test cases
2. Telemetry and Diagnostic logging.

### CODE REVIEW

File: src/Agent.Plugins/Loggers/TraceLogger.cs [hunk 1]
```diff
@@ -20,13 +20,13 @@ public static ITraceLogger Instance
...
         public void Error(string error)
         {
-            throw new NotImplementedException();
+            //throw new NotImplementedException();
         }
 
         /// <inheritdoc />
         public void Info(string text)
         {
-            throw new NotImplementedException();
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Loggers/TraceLogger.cs]: revert

File: src/Agent.Plugins/Parser/Python/PythonRegularExpressions.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿namespace Agent.Plugins.TestResultParser.Parser.Python
+{
+    using System.Text.RegularExpressions;
+
+    public class PythonRegularExpressions
+    {
+        public static readonly Regex ResultPattern = new Regex($"^(?<{RegexCaptureGroups.TestCaseName}>.+) \\.\\.\\. (?<{RegexCaptureGroups.TestOutcome}>.*)$", RegexOptions.ExplicitCapture);
+
+        // TODO: Have separate pattern for error if required
+        public static readonly Regex FailedResultPattern = new Regex($"^(FAIL|ERROR)( )?:(?<{RegexCaptureGroups.TestCaseName}>(.+))$", RegexOptions.ExplicitCapture);
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonRegularExpressions.cs]: >CaptureGroups.TestCaseNam [](start = 98, length = 25)

include ( )? after :

File: src/Agent.Plugins/Parser/Python/PythonRegularExpressions.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+        public static readonly Regex PassedOutcomePattern = new Regex(@"(^ok|expected failure)|( (ok|expected failure))$");
+        public static readonly Regex SkippedOutcomePattern = new Regex(@"^skipped");
+
+        public static readonly Regex SummaryTestCountAndTimePattern = new Regex(@"^Ran ([0-9]+) tests? in ([0-9]+)(\.([0-9]+))?s");
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonRegularExpressions.cs]: $

File: src/Agent.Plugins/Parser/Python/PythonRegularExpressions.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        public static readonly Regex SummaryAllowedWhiteSpaceLine = new Regex(@"^((\s|\t)+)?$");
+        public static readonly Regex SummaryOutcomePattern = new Regex(@"(OK|FAILED)( \((.*)\))?");
+        public static readonly Regex SummaryFailurePattern = new Regex($"(^|, )failures( )?=(?<{RegexCaptureGroups.FailedTests}>[1-9][0-9]*)", RegexOptions.ExplicitCapture);
+        public static readonly Regex SummaryErrorsPattern = new Regex($"(^|, )errors( )?=(?<{RegexCaptureGroups.Errors}>[1-9][0-9]*)", RegexOptions.ExplicitCapture);
+        public static readonly Regex SummarySkippedPattern = new Regex($"(^|, )skipped( )?=(?<{RegexCaptureGroups.SkippedTests}>[1-9][0-9]*)", RegexOptions.ExplicitCapture);
+    }
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonRegularExpressions.cs]: try and put in a single regex
actual pattern as a comment

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 5]
```diff
@@ -1,42 +1,295 @@
...
-        {
+        /// <summary>
+        /// Detailed constructor where specified logger and telemetry data collector are initialized along with test run manager
+        /// </summary>
+        /// <param name="testRunPublisher"></param>
+        /// <param name="diagnosticDataCollector"></param>
+        /// <param name="telemetryDataCollector"></param>
+        public PythonTestResultParser(ITestRunManager testRunManager, ITraceLogger traceLogger, ITelemetryDataCollector telemetryCollector)
+        {
+            traceLogger.Info("MochaTestResultParser.MochaTestResultParser : Starting mocha test result parser.");
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: >mocha [](start = 85, length = 5)

python

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 6]
```diff
@@ -1,42 +1,295 @@
...
-            throw new NotImplementedException();
-        }
+        /// <param name="logData">Data to be parsed.</param>
+        public void Parse(LogData logData)
+        {
+            // Validate data input
+            if (!this.IsValidInput(logData.Line)) return;
+
+            // Switch to proper parser state
+            this.logger.Verbose($"Current state: {nameof(this.state)}");
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: not necessary, only on transitions

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 7]
```diff
@@ -1,42 +1,295 @@
...
+                    if (ParseSummary(logData))
+                    {
+                        state = ParserState.ExpectingSummary;
+                        return;
+                    }
+
+                    if (ParseTestResult(logData)) return;
+                    break;
+
+                default:
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: >default [](start = 16, length = 7)

add actual case

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 8]
```diff
@@ -1,42 +1,295 @@
...
+                    if (ParseTestResult(logData)) return;
+                    if (ParseForFailedResult(logData))
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: >ParseForFailedResult [](start = 24, length = 20)

add comments explaining logic

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 9]
```diff
@@ -1,42 +1,295 @@
...
+                    {
+                        partialTestResult = null;
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: comment

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 10]
```diff
@@ -1,42 +1,295 @@
...
+            // Validate data input
+            if (!this.IsValidInput(logData.Line)) return;
+
+            // Switch to proper parser state
+            this.logger.Verbose($"Current state: {nameof(this.state)}");
+
+            switch (state)
+            {
+                case ParserState.ExpectingSummary:
+                    if (ParseSummary(logData)) return;
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: >ParseSummary [](start = 24, length = 12)

tryparse, trymatch

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 11]
```diff
@@ -1,42 +1,295 @@
...
+            {
+                var partialResultMatch = PythonRegularExpressions.PassedOutcomePattern.Match(logData.Line);
+                if (partialResultMatch.Success)
+                {
+                    partialTestResult.Outcome = TestOutcome.Passed;
+                    currentTestRun.PassedTests.Add(partialTestResult);
+                }
+                return false;
+            }
+            else
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: >else [](start = 12, length = 4)

not required

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 12]
```diff
@@ -1,42 +1,295 @@
...
+            {
+                partialTestResult = null;
+            }
+
+            // Test result name
+            var resultNameIdentifier = resultMatch.Groups[RegexCaptureGroups.TestCaseName].Value.Trim();
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: >.Value.Trim(); [](start = 90, length = 14)

try doing this in the regex

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 13]
```diff
@@ -1,42 +1,295 @@
...
+            if (currentTestRun.TestRunSummary == null)
+            {
+                var countAndTimeSummaryMatch = PythonRegularExpressions.SummaryTestCountAndTimePattern.Match(logData.Line);
+
+                if (countAndTimeSummaryMatch.Success)
+                {
+                    var testcount = int.Parse(countAndTimeSummaryMatch.Groups[1].Value);
+                    var secTime = int.Parse(countAndTimeSummaryMatch.Groups[2].Value);
+                    var msTime = int.Parse(countAndTimeSummaryMatch.Groups[4].Value);
+
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: use capture groups

File: src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs [hunk 14]
```diff
@@ -1,42 +1,295 @@
...
+                    currentTestRun.TestRunSummary = new TestRunSummary();
+                    currentTestRun.TestRunSummary.TotalExecutionTime = new TimeSpan(0, 0, 0, secTime, msTime);
+                    currentTestRun.TestRunSummary.TotalTests = testcount;
+                    return true;
+                }
+            }
+            else
```
Comment (ShreyasRmsft) [inline src/Agent.Plugins/Parser/Python/PythonTestResultParser.cs]: >else [](start = 12, length = 4)

remove else and add return above

**Comment by singhsarab** [comment]
@ShreyasRmsft Reacted to all your comments. Pushing this in, if you have any comments, I'll take them as part of next PR.

**Pull request merged**